app/scripts/lib/freezePromise.js

```diff
@@ -0,0 +1,16 @@
+
+/**
+ * Freezes the Promise global and prevents its reassignment.
+ */
+const deepFreeze = require('deep-freeze-strict')
+
+if (
+  process.env.IN_TEST !== 'true' &&
+  process.env.METAMASK_ENV !== 'test'
+) {
+  Promise = deepFreeze(Promise) /* eslint-disable-line */
+  const { value, enumerable } = Object.getOwnPropertyDescriptor(global, 'Promise')
+  Object.defineProperty(global, 'Promise', {
+    value, enumerable, configurable: false, writable: false,
```

Comment:
This can be simplified somewhat; `value` and `enumerable` can be omitted, meaning that the `getOwnPropertyDescriptor` call isn't needed.

Reply:
Hrm, MDN would lead me to believe that neither can be omitted: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty

Surely the `value` must be specified, even if we just do `value: Promise` directly?

MDN would also lead us to believe that `configurable` and `writable` can be omitted but that is not in fact the case. They appear to default to `true`.

Reply:
In practice they do not seem to be necessary; at least not in the versions of Chrome and Firefox I'm using. They are all described as optional in that MDN article as well.

The defaults listed in the article are used for _new_ properties defined with `defineProperty`, not existing properties.

Reply:
in node v10.16.0, those properties defaulted to `true` to my surprise

Reply:
id wrap this in a function that freezes a key on a target. you'll likely want to freeze other things in the future

Reply:
I thoroughly investigated this and truth appears to be somewhere in-between. Calling `Object.defineProperty` using node v10.16.0 and running `yarn start` on my machine:
- without `configurable: false` the property can be deleted
- without `writable: false` the property can be overwritten
- `value` and `enumerable` appear to use what the property in question already uses if omitted

Reply:
All requested changes have been addressed except what's mentioned in my previous comment.

Reply:
I think there's been some confusion here; it does default to `false` on Node v10.16.0. But the defaults only apply if the descriptor hasn't been set yet - if it doesn't _already_ have a value (i.e. new properties).

e.g.:
```
$ node
> const foo = { prop1: {} }
undefined
> Object.defineProperty(foo, 'prop1', {})
{ prop1: {} }
> Object.getOwnPropertyDescriptor(foo, 'prop1')
{ value: {},
  writable: true,
  enumerable: true,
  configurable: true }
> Object.defineProperty(foo, 'prop2', {})
{ prop1: {} }
> Object.getOwnPropertyDescriptor(foo, 'prop2')
{ value: undefined,
  writable: false,
  enumerable: false,
  configurable: false }
```

As we can be confident that `Promise` will always be declared already, so the "defaults" for the `defineProperty` descriptor are not relevant to this case. `Object.defineProperty` will change whatever we request that it change, and leave the rest alone.

Reply:
Ah, so if a property is already there, it just uses whatever is returned from `getOwnPropertyDescriptor` for all values except those specified?

Reply:
Yep, exactly.

It's a bit confusing using the term "defaults" because it makes it sound like it means a default for the `defineProperty` operation itself, but really it's a default for the underlying property. So it only gets set if nothing's set there already.